src/events/input_provider.rs

```diff
@@ -0,0 +1,254 @@
+//! Contains the `InputProvider` trait, which is used to provide input events to widgets.
+
+use events::{UiEvent, Scroll, MouseClick, MouseDrag, InputState};
+use input::{Input, Button};
+use input::keyboard::Key;
+use input::mouse::MouseButton;
+use position::Point;
+
+
+/// Trait for something that provides events to be consumed by a widget.
+/// Provides a bunch of convenience methods for filtering out specific types of events.
+pub trait InputProvider<'a, T: Iterator<Item=&'a UiEvent>> {
```

Comment:
I think it would be a little more idiomatic to move this `T` type parameter to an associated type, i.e. something like `type Events: Iterator<Item=&'a UiEvent>`. This is because we know that there may only be one type of event iterator per implementation of `InputProvider`, i.e. `GlobalInput`'s is `GlobalInputEventIterator` and `WidgetInput`'s is `WidgetInputEventIterator`.